lib/browser/http.ts

```diff
@@ -193,3 +194,109 @@ export class HttpClientStream extends BufferedEventEmitter {
         this.emit('error', new Error(`msg=${error.message}, XHR=${error.request}, info=${info}`));
     }
 }
+
+interface PendingRequest {
+    resolve: (connection: HttpClientConnection) => void;
+    reject: (error: CrtError) => void;
+}
+
+/** Creates, manages, and vends connections to a given host/port endpoint */
+export class HttpClientConnectionManager {
+    private pending_connections = new Set<HttpClientConnection>();
+    private live_connections = new Set<HttpClientConnection>();
+    private free_connections: HttpClientConnection[] = [];
+    private pending_requests: PendingRequest[] = [];
+
+    constructor(
+        readonly host: string,
+        readonly port: number,
+        readonly max_connections: number
+    ) {
+        
+    }
+
+    private remove(connection: HttpClientConnection) {
+        this.pending_connections.delete(connection);
+        this.live_connections.delete(connection);
+        const free_idx = this.free_connections.indexOf(connection);
+        if (free_idx != -1) {
+            this.free_connections.splice(free_idx, 1);
+        }
+    }
+
+    private resolve(connection: HttpClientConnection) {
+        const request = this.pending_requests.shift();
+        if (request) {
+            request.resolve(connection);
+        }
+    }
+
+    private reject(error: CrtError) {
+        const request = this.pending_requests.shift();
+        if (request) {
+            request.reject(error);
+        }
+    }
+
+    private pump() {
+        // Try to service the request with a free connection
+        {
```

Comment:
There are certain circumstances where pump gets called when there are no pending requests in which case it doesn't seem like the free list should get popped or a new connection created.